autoload/ale/go.vim

```diff
@@ -0,0 +1,23 @@
+" Author: Horacio Sanson https://github.com/hsanson
+" Description: Functions for integrating with Go tools
+
+" Find the nearest dir listed in GOPATH and assume it the root of the go
+" project.
+function! ale#go#FindProjectRoot(buffer) abort
+
+    let l:sep = has('win32') ? ';' : ':'
+
+    let l:filename = ale#path#Simplify(expand('#' . a:buffer . ':p'))
+
+    for l:name in split($GOPATH, l:sep)
```

Comment:
`$GOPATH` may be empty; from `go help gopath`:

> If the environment variable is unset, GOPATH defaults
to a subdirectory named "go" in the user's home directory
($HOME/go on Unix, %USERPROFILE%\go on Windows),
unless that directory holds a Go distribution.
Run "go env GOPATH" to see the current GOPATH.

Reply:
Added code to return $HOME/go on Unix or  %USERPROFILE%\go on Windows if  they exists and GOPATH is not defined.

Reply:
I don't know what it does exactly, but it seems to me that if you're going to set it, you might as well set it to a correct value, and not a surprising incorrect one?

Reply:
It is setting the value based exactly on what `go help gopath` says it should be. So this should be the correct path for any environment:

 - 1. first try to find a root defined in GOPATH
 - 2. If GOPATH is not defined then set it to $HOME/go on Unix or %USERPROFILE%\go on Windows.
 - 3 If none of the above are defined or not exist then returns empty string, that disables the linter.

Reply:
But it's not the project's directory, it's the GOPATH?

Reply:
As I mentioned on a separate conversation the root path is irrelevant for go-langserver. The cases were we want the actual root path is for language servers that support special configuration files (e.g. .solargraph.yml for ruby language server). In this case we can search for these special configuration files and assume their directory as root project path.

The go-langserver does not support any configuration files so I avoid the trouble of finding the actual root path. Getting the actual root path of a go project is difficult (except for go modules that have a go.mod file) and has no actual use for go-langserver.

Reply:
Yes, I read that but "it just happens to work" is no reason to do the wrong thing. When it *does* matter in 6 months or 2 years everyone will be confused. Or people looking for this "get project root" utility function for something else will be confused.

It would be much better to just not do anything (e.g. set it to the current dir) than to a different value than you're telling people to set it.